tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift

```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+                                 to: LPTHREAD_START_ROUTINE.self)
+    let hThread: HANDLE = CreateRemoteThread(self.process, nil, 0, hookAddr,
+                                             dllPathRemote, 0, nil)
+    if hThread == HANDLE(bitPattern: 0) {
+      print("CreateRemoteThread failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hThread) }
+
+    // The main heap iteration loop.
+    while true {
+      let wait = WaitForSingleObject(hReadEvent, waitTimeoutMs)
+      if wait != WAIT_OBJECT_0 {
+        print("WaitForSingleObject failed \(wait)")
+        return
       }
 
-      repeat {
-        if entry.dwFlags & DWORD(LF32_FREE) == DWORD(LF32_FREE) { continue }
-        body(swift_addr_t(entry.dwAddress), UInt64(entry.dwBlockSize))
-      } while Heap32Next(&entry)
+      let entryCount = bufSize/MemoryLayout<HeapEntry>.size
+      let buffer = buf.bindMemory(to: HeapEntry.self, capacity: 1)
+      let entry = buffer.pointee
+      if entry.Address == UInt.max {
+        // The buffer containing zero entries, indicated by the first entry
+        // contains -1, means, we are done. Break out of loop.
+        if !SetEvent(hWriteEvent) {
+          print("SetEvent failed \(GetLastError())")
+          return
+        }
+        break
+      }
+      for i in 0..<entryCount {
+        let entry = buffer.advanced(by: i).pointee
+        if entry.Address == 0 {
+          // Done. Break out of loop.
+          break
+        }
+        body(swift_addr_t(entry.Address), UInt64(entry.Size))
+      }
 
-      heap.dwSize = SIZE_T(MemoryLayout<HEAPLIST32>.size)
-    } while Heap32ListNext(hSnapshot, &heap)
+      if !SetEvent(hWriteEvent) {
+        print("SetEvent failed \(GetLastError())")
+        return
+      }
+    }
+
+    let wait = WaitForSingleObject(hThread, waitTimeoutMs)
+    if wait != WAIT_OBJECT_0 {
+      print("WaitForSingleObject on LoadLibrary failed \(wait)")
+      return
+    }
+
+    var threadExitCode: DWORD = 0
+    GetExitCodeThread(hThread, &threadExitCode)
+    if threadExitCode == 0 {
+      print("LoadLibraryW failed \(threadExitCode)")
+      return
+    }
+
+    // Unload the dll and deallocate the dll path from the remote process
+    if !unloadDllAndPathRemote(dwProcessId: dwProcessId, hKernel32: hKernel32,
+                               dllPathRemote: dllPathRemote) {
+      print("Failed to unload the remote dll")
+      return
+    }
+  }
+
+  private func allocateDllPathRemote() -> UnsafeMutableRawPointer? {
+    // The path to the dll assuming it's in the same directory as swift-inspect.
+    let swiftInspectPath = ProcessInfo.processInfo.arguments[0]
+    return URL(fileURLWithPath: swiftInspectPath)
+        .deletingLastPathComponent()
+        .appendingPathComponent("SwiftInspectClient.dll")
+        .withUnsafeFileSystemRepresentation {
+      #"\\?\\#(String(decodingCString: unsafeBitCast($0!, to: UnsafePointer<UInt8>.self), as: UTF8.self))"#
```

Comment:
Why is the `unsafeBitCast` needed here?

Reply:
A cast from `UnsafePointer<Int8>` to `UnsafePointer<UInt8>` needed here

Reply:
That doesn't seem to work. 

```
S:\SourceCache\swift\tools\swift-inspect\Sources\swift-inspect\WindowsRemoteProcess.swift:421:45: error: value of type 'UnsafePointer<Int8>?' has no member 'baseAddress'
        #"\\?\\#(String(decodingCString: $0.baseAddress, as: UTF8.self))"#
                                         ~~ ^~~~~~~~~~~
```
```
S:\SourceCache\swift\tools\swift-inspect\Sources\swift-inspect\WindowsRemoteProcess.swift:421:46: error: value of type 'UnsafePointer<Int8>' has no member 'baseAddress'
        #"\\?\\#(String(decodingCString: $0!.baseAddress, as: UTF8.self))"#
```